File: src/main/kotlin/merge/Runner.kt
package com.revtekk.blaze.merge

import com.revtekk.blaze.common.Command
import com.revtekk.blaze.common.readLines

class MergeCommand: Command {
    constructor(args: List<String>) : super(args) {
        if (args.size != 3) {
            throw IllegalArgumentException("merge: insufficient arguments specified, expected 3, got ${args.size}")
        }
    }

    override fun execute() {
        /*
         * original => the common ancestor of v1 and v2
         * v1 => modification of original
         * v2 => modification of original
         *
         * objective: merge v1, v2 using the original as a guide (aka 3-way merge)
         */
        val original = args[0]
        val v1 = args[1]
        val v2 = args[2]

        val originalLines = readLines(original)
        val v1Lines = readLines(v1)
        val v2Lines = readLines(v2)

        val result = merge(originalLines, v1Lines, v2Lines)
        result.segments.forEach {
            // FIXME once merge conflict detection is implemented, need to handle this
            if (it is MergeLines) {
                println(it.lines.joinToString("\n"))
            }
        }
    }
}

File: src/merge/runner.rs
use crate::command::{Command, CommandExecutionError};
use crate::file::read_lines;
use crate::merge::{merge, MergeSegmentType};

pub struct MergeCommand {
    pub args: Vec<String>,
}

impl Command for MergeCommand {
    fn execute(&self) -> Result<(), CommandExecutionError> {
        if self.args.len() != 3 {
            return Err(CommandExecutionError {
                message: format!(
                    "merge: insufficient arguments specified, expected 3, got {size}",
                    size = self.args.len()
                ),
            });
        }

        /*
         * original => the common ancestor of v1 and v2
         * v1 => modification of original
         * v2 => modification of original
         *
         * objective: merge v1, v2 using the original as a guide (aka 3-way merge)
         */
        let original = &self.args[0];
        let v1 = &self.args[1];
        let v2 = &self.args[2];

        let original_lines = match read_lines(original) {
            Ok(lines) => lines,
            Err(e) => return Err(e),
        };
        let v1_lines = match read_lines(v1) {
            Ok(lines) => lines,
            Err(e) => return Err(e),
        };
        let v2_lines = match read_lines(v2) {
            Ok(lines) => lines,
            Err(e) => return Err(e),
        };

        let result = merge(&original_lines, &v1_lines, &v2_lines);

        for segment in result.segments {
            match segment.segment_type {
                MergeSegmentType::Lines => {
                    println!("{}", segment.lines.join("\n"));
                }
                MergeSegmentType::Conflict => {
                    println!("--- @blaze:mg_conflict:v1 ---");
                    println!("{}", segment.v1_changes.join("\n"));
                    println!("--- @blaze:mg_conflict:v2 ---");
                    println!("{}", segment.v2_changes.join("\n"));
                    println!("--- @blaze:mg_conflict:end ---");
                }
            }
        }

        Ok(())
    }
}
